sql: reset conn buffers after each query

DrewKimball opened the pull request:
**sql: reset conn buffers after each query**

There are two buffers in `pgwire.conn` that are used when sending results
to the client. They have to buffer each row in its entirety before flushing,
so wide rows can cause the capacity of these buffers to exceed the limit.

Previously, the buffers were never reallocated even when they exceeded the
size limit. This could cause a long-running session to hold on to large
amounts of memory until the session was ended. This increased the risk of
OOM, in addition to being inefficient.

This commit adds a method `maybeReallocate` to `pgwire.conn`, which
reallocates each buffer if it has reached the size limit. `maybeReallocate`
is called upon closing `Sync` and `Flush` command results. These are
natural places to reallocate the buffers, since they already flush the
buffers to the client and are called roughly at the statement granularity,
rather than for every row. This allows long-running sessions to bound their
memory usage without causing reallocation on every wide row that is sent
to the client.

Informs #80497

Release note (performance improvement): Long-running sql sessions are
now less likely to maintain large allocations for long periods of time,
which decreases the risk of OOM and improves memory utilization.

Release justification: bug fix to reduce likelihood of OOM

**sql: remove unnecessary maybeFlush call**

This commit fixes an oversight of #83870 that added a redundant call to
`maybeAppend` in `limitedCommandResult.AddRow`.

Release note: None

Release justification: low-risk fix for oversight in previous bug fix

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85949)
<!-- Reviewable:end -->

DrewKimball commented:
To verify that this change works as expected, I ran the following queries in the same session on a single demo node:
```
--Return a wide row to the client to grow the buffers.
SELECT string_agg(x, '')::BYTES FROM (SELECT 'vbX5ajNEOh1VlLJ4' FROM generate_series(1, 400000)) s(x);

--Sleep to simulate a long-running query that keeps the session active.
SELECT pg_sleep(600);
```
I then waited 3 minutes to ensure the first query was finished and that garbage collection had time to take place. Here are the resulting memory profiles for the node:
**Master**
<img width="1566" alt="Screen Shot 2022-08-11 at 1 40 40 AM" src="https://user-images.githubusercontent.com/46886010/184103658-fde5a8c3-6fce-4647-92fd-edcc8e694ab3.png">
**Branch**
<img width="1561" alt="Screen Shot 2022-08-10 at 7 31 26 PM" src="https://user-images.githubusercontent.com/46886010/184103674-8a10b9c4-6c63-416a-9785-19f6b1b3995e.png">

knz reviewed (commented):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/pgwire/command_result.go` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85949#-N9BFPij-KvDSugIi1lD:-N9BFPij-KvDSugIi1lE:bjxlu1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ae7ebe2145839f7559c824a0a58f85017fd7fe7/pkg/sql/pgwire/command_result.go#L440)):*
> ```Go
> 		return r.moreResultsNeeded(ctx)
> 	}
> 	return nil
> ```

The calls to maybeReallocate above are fine, but are you sure you want to remove the maybeFlush call here? 


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/pgwire/command_result.go` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85949#-N9BFPij-KvDSugIi1lD:-N9BJ6TZ7gxcYi_Kh238:brta0sd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ae7ebe2145839f7559c824a0a58f85017fd7fe7/pkg/sql/pgwire/command_result.go#L440)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The calls to maybeReallocate above are fine, but are you sure you want to remove the maybeFlush call here? 
</blockquote></details>

Thanks for taking a look. I added the call in #83870, but @mgartner pointed out it isn't necessary because the call to `commandResult.AddRow` above already calls `maybeFlush`.


<!-- Sent from Reviewable.io -->

knz commented:
kk

DrewKimball commented on pkg/sql/pgwire/conn.go:
```diff
@@ -1599,6 +1599,20 @@ func (c *conn) maybeFlush(pos sql.CmdPos, bufferingDisabled bool) error {
+// maybeReallocate should only be called after the connection has been flushed
+// and a query has just been processed.
+func (c *conn) maybeReallocate() {
+	limit := int(c.sessionArgs.ConnResultsBufferSize)
```
It looks like that argument can be set when the connection is created, but can't be changed once the session is active: https://github.com/cockroachdb/cockroach/blob/5845f4bc637bff4595d9ad9feb5fbaade875559a/pkg/sql/vars.go#L1015-L1021
I added some comments to make this more clear.

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85949#-:-N9DVy6i44AxOk4hpwxf:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @knz, @mgartner, and @yuzefovich)

___
*[`pkg/sql/pgwire/conn.go` line 1607 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85949#-N9BCdEMqgT-eWjY9O4r-r1-1607:-N9DVwXR4Ua4Ltg8A_8T:b-bl1gfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ae7ebe2145839f7559c824a0a58f85017fd7fe7/pkg/sql/pgwire/conn.go#L1607)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It looks like that argument can be set when the connection is created, but can't be changed once the session is active: https://github.com/cockroachdb/cockroach/blob/5845f4bc637bff4595d9ad9feb5fbaade875559a/pkg/sql/vars.go#L1015-L1021
I added some comments to make this more clear.
</blockquote></details>

thanks for explaining


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85949#-:-N9Xe-iQ03N_xB4FJ8hP:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/pgwire/command_result.go` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85949#-N9BFPij-KvDSugIi1lD:-N9XD6fh-hTld1Jn8VDq:bvws23i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ae7ebe2145839f7559c824a0a58f85017fd7fe7/pkg/sql/pgwire/command_result.go#L440)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Thanks for taking a look. I added the call in #83870, but @mgartner pointed out it isn't necessary because the call to `commandResult.AddRow` above already calls `maybeFlush`.
</blockquote></details>

It'd be cleaner to pull this change out into a separate commit since it fixes a minor oversight of #83870.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: reset conn buffers after each query

There are two buffers in `pgwire.conn` that are used when sending results
to the client. They have to buffer each row in its entirety before flushing,
so wide rows can cause the capacity of these buffers to exceed the limit.

Previously, the buffers were never reallocated even when they exceeded the
size limit. This could cause a long-running session to hold on to large
amounts of memory until the session was ended. This increased the risk of
OOM, in addition to being inefficient.

This commit adds a method `maybeReallocate` to `pgwire.conn`, which
reallocates each buffer if it has reached the size limit. `maybeReallocate`
is called upon closing `Sync` and `Flush` command results. These are
natural places to reallocate the buffers, since they already flush the
buffers to the client and are called roughly at the statement granularity,
rather than for every row. This allows long-running sessions to bound their
memory usage without causing reallocation on every wide row that is sent
to the client.

Informs #80497

Release note (performance improvement): Long-running sql sessions are
now less likely to maintain large allocations for long periods of time,
which decreases the risk of OOM and improves memory utilization.

Release justification: bug fix to reduce likelihood of OOM

DrewKimball committed:
sql: remove unnecessary maybeFlush call

This commit fixes an oversight of #83870 that added a redundant call to
`maybeAppend` in `limitedCommandResult.AddRow`.

Release note: None

Release justification: low-risk fix for oversight in previous bug fix

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85949)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @knz, @mgartner, and @yuzefovich)

___
*[`pkg/sql/pgwire/command_result.go` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85949#-N9BFPij-KvDSugIi1lD:-N9Xl2kL5grYZ-JfQ7rD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ae7ebe2145839f7559c824a0a58f85017fd7fe7/pkg/sql/pgwire/command_result.go#L440)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It'd be cleaner to pull this change out into a separate commit since it fixes a minor oversight of #83870.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

DrewKimball commented:
Test failures look like unrelated flakes.

bors r+

mgartner commented:
I think we should backport this. Do you all agree?

yuzefovich commented:
I'm in favor.

mgartner commented:
👍 @DrewKimball Can you backport this to 22.1 and 21.2?

DrewKimball commented:
Will do

DrewKimball commented:
blathers backport 22.1 21.2

mgartner commented:
Oh lol... I forgot blathers could do that - I would have done it myself. Sorry!

DrewKimball commented:
To be honest, I wasn't expecting it to be so smooth. I'm pleasantly surprised
